src/data-sources/plugins/CosmosDB/Query.ts

```diff
@@ -0,0 +1,139 @@
+import * as request from 'xhr-request';
+import { DataSourcePlugin, IOptions } from '../DataSourcePlugin';
+import { DataSourceConnector } from '../../DataSourceConnector';
+import CosmosDBConnection from '../../connections/cosmos-db';
+
+let connectionType = new CosmosDBConnection();
+
+interface IQueryParams {
+  query?: ((dependencies: any) => string) | string;
+  parameters?: (string | object)[];
+  mappings?: (string | object)[];
+  databaseId?: string;
+  collectionId?: string;
+  calculated?: (results: any) => object;
+}
+
+export default class CosmosDBQuery extends DataSourcePlugin<IQueryParams> {
+  type = 'CosmosDB-Query';
+  defaultProperty = 'values';
+  connectionType = connectionType.type;
+
+  constructor(options: IOptions<IQueryParams>, connections: IDict<IStringDictionary>) {
+    super(options, connections);
+    this.validateTimespan(this._props);
+    this.validateParams(this._props.params);
+  }
+
+  /**
+   * update - called when dependencies are created
+   * @param {object} dependencies
+   * @param {function} callback
+   */
+  updateDependencies(dependencies: any) {
+    let emptyDependency = false;
+    Object.keys(this._props.dependencies).forEach((key) => {
+      if (typeof dependencies[key] === 'undefined') { emptyDependency = true; }
+    });
+
+    // If one of the dependencies is not supplied, do not run the query
+    if (emptyDependency) {
+      return (dispatch) => {
+        return dispatch();
+      };
+    }
+
+    // Validate connection
+    let connection = this.getConnection();
+    let { host, key } = connection;
+    if (!connection || !host || !key) {
+      return (dispatch) => {
+        return dispatch();
+      };
+    }
+
+    // Get Dashboard Id
+    const paths = location.pathname.split('/');
+    if (paths.length !== 3) {
+      throw Error('Expected location pathname:' + paths);
+    }
+    const dashboardId = paths[paths.length - 1];
+
+    const params = this._props.params;
+    const query: string = this.compileQuery(params.query, dependencies);
+
+    const url = `/cosmosdb/dashboard/${dashboardId}`;
```

Comment:
I think you can change this to be a url that receives all of its parameters in the post parameters (no need for the server to poll the parameters from the dashboard itself)

We can discuss this to see what makes sense.